### PR TITLE
Generate: improve docstrings for custom stopping criteria

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Addresses https://github.com/huggingface/transformers/issues/18068, and a few other GH issues with the same problem

When a user defines a custom stopping criteria that makes use of the `scores` input, if `return_dict_in_generate=True, output_scores=True` are not passed, an exception will be thrown (`scores` is `None`). This is because we only keep track of `scores` when those two flags are passed.

I've added this requirement in several places in the docstrings, to reduce the odds of being missed.

For completeness: I've also considered always keeping track of scores. However, this adds a non-negligible overhead -- in BF16 for the 7B Mistral model, we would need an extra ~50MB per 1000 generated tokens (and batching would act as a multiplier here). Ditto for 4-bit BNB. Since custom stopping criteria are seldom used, this didn't look like a good tradeoff :)

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
Sounds, good! Would be down to raise an error for logits processors that leverage scores when `return_dict_in_generate=True,output_scores=True` is not set but they don't have access to it so too cumbersome!
